lib/kdtree.h

```diff
@@ -267,83 +153,103 @@ class FlatNode {
-            return !(*this == other);
-        }
-
-        friend struct std::hash<OptionalId>;
+    explicit KDTree(Triangles tris);
 
-    private:
-        TriangleId id_;
-    };
+    size_t height() const {
+        using Node = detail::FlatNode;
+        std::stack<std::pair<const Node*, uint32_t /* level */>> stack;
+        const Node* root = nodes_.data();
+        stack.emplace(root, 0);
+        size_t height = 0;
+        while (!stack.empty()) {
+            size_t level = stack.top().second;
+            if (level > height) {
+                height = level;
+            }
 
-    explicit KDTree(Triangles tris);
+            const Node* node = stack.top().first;
+            stack.pop();
 
-    size_t height() const { return root_->height(); }
-    size_t size() const { return root_->size(); }
+            if (node->is_inner()) {
+                stack.emplace(node + 1, level + 1);
+                stack.emplace(root + node->right(), level + 1);
+            }
+        }
+        return height;
+    }
+    size_t num_nodes() const { return nodes_.size(); }
     size_t num_triangles() const { return tris_.size(); }
     const Triangles& triangles() const { return tris_; }
     const Box& box() const { return box_; }
     const Triangle& operator[](const TriangleId id) const { return tris_[id]; }
     const Triangle& at(const TriangleId id) const { return tris_.at(id); }
 
-    static constexpr size_t node_size() { return sizeof(Node); }
-
-private:
-    // Helper method for recursive construction
-    Node* build(TriangleIds tris, const Box& box);
-
-    // Cf. [WH06], Algorithm 4
-    std::tuple<float /*cost*/, Axis /* plane axis */, float /* plane pos */,
-               TriangleIds /*left*/, TriangleIds /*right*/>
-    find_plane_and_classify(const TriangleIds& tris, const Box& box) const;
-
-    void flatten();
+    static constexpr size_t node_size() { return sizeof(detail::FlatNode); }
 
 private:
-    std::unique_ptr<Node> root_;
     Triangles tris_;
     Box box_;
 
-    // TODO: Describe layout
-    std::vector<detail::FlatNode> nodes_; // flatten nodes
+    /**
+     * Layout:
+     *
+     * We have 2 types of nodes, cf. Node: inner nodes and leaf nodes. All
+     * nodes are stored in DFS order. Next neighbor node of a node is always its
+     * left child, to the right child the node stores an index. A leaf node
+     * contains 2 or 1 valid references into triangles_ array. If an inner node
+     * points to a leaf node, then all consecutive leaf nodes of the first one
+     * describe the set of triangles belonging to the leaf node. The last node
+     * in the array is an inner node (sentinel) and does not contain any other
+     * valid data.
+     *
+     * Cf. possible layout (numbers are indexes in the array):
+     *
+     *         0
+     *        / \
+     *       /   \
+     *      1     7 8/sentinel
+     *     / \
+     *    /   \
+     *   2 3   4 5 6
+     */
```

Comment:
I don't quite get this. We should write out the array with the flat node values. So

```
nodes_[0] == int(1), int(7), bits(1,1)
```
Is that correct?

Whats's the value of `nodes_[4]`, `nodes_[5]` and `nodes_[6]`?